repo/content/index_blob_manager.go

```diff
@@ -0,0 +1,325 @@
+	var hashOutput [maxHashSize]byte
+
+	hash := m.hasher(hashOutput[:0], data)
+	blobID := prefix + blob.ID(hex.EncodeToString(hash))
+
+	iv, err := getIndexBlobIV(blobID)
+	if err != nil {
+		return blob.Metadata{}, err
+	}
+
+	data2, err := m.encryptor.Encrypt(nil, data, iv)
+	if err != nil {
+		return blob.Metadata{}, err
+	}
+
+	m.listCache.deleteListCache(prefix)
+
+	err = m.st.PutBlob(ctx, blobID, gather.FromSlice(data2))
+	if err != nil {
+		return blob.Metadata{}, err
+	}
+
+	bm, err := m.st.GetMetadata(ctx, blobID)
+	if err != nil {
+		return blob.Metadata{}, errors.Wrap(err, "unable to get blob metadata")
+	}
+
+	if err := m.ownWritesCache.add(ctx, bm); err != nil {
+		log(ctx).Warningf("unable to cache own write: %v", err)
+	}
+
+	return bm, nil
+}
+
+func (m *indexBlobManagerImpl) getCompactionLogEntries(ctx context.Context, blobs []blob.Metadata) (map[blob.ID]*compactionLogEntry, error) {
+	results := map[blob.ID]*compactionLogEntry{}
+
+	for _, cb := range blobs {
+		data, err := m.getIndexBlob(ctx, cb.BlobID)
+
+		if errors.Is(err, blob.ErrBlobNotFound) {
+			continue
+		}
+
+		if err != nil {
+			return nil, errors.Wrapf(err, "unable to read compaction blob %q", cb.BlobID)
+		}
+
+		le := &compactionLogEntry{}
+
+		if err := json.Unmarshal(data, le); err != nil {
+			return nil, errors.Wrap(err, "unable to read compaction log entry %q")
+		}
+
+		results[cb.BlobID] = le
+	}
+
+	return results, nil
+}
+
+func (m *indexBlobManagerImpl) deleteOldIndexBlobs(ctx context.Context, latestBlob blob.Metadata) error {
+	if m.minCompactionLogBlobDeleteAge <= m.minIndexBlobDeleteAge {
+		return errors.Errorf("configuration error - compaction log deletion age (%v) must be greater than index blob deletion age (%v)", m.minCompactionLogBlobDeleteAge, m.minIndexBlobDeleteAge)
+	}
+
+	allCompactionLogBlobs, err := m.listCache.listIndexBlobs(ctx, compactionLogBlobPrefix)
+	if err != nil {
+		return errors.Wrap(err, "error listing compaction log blobs")
+	}
+
+	// look for server-assigned timestamp of the compaction log entry we just wrote as a reference.
+	// we're assuming server-generated timestamps are somewhat reasonable and time is moving
+	compactionLogServerTimeCutoff := latestBlob.Timestamp.Add(-m.minIndexBlobDeleteAge)
+	compactionBlobs := blobsOlderThan(allCompactionLogBlobs, compactionLogServerTimeCutoff)
+
+	log(ctx).Debugf("fetching %v/%v compaction logs older than %v", len(compactionBlobs), len(allCompactionLogBlobs), compactionLogServerTimeCutoff)
+
+	compactionBlobEntries, err := m.getCompactionLogEntries(ctx, compactionBlobs)
+	if err != nil {
+		return errors.Wrap(err, "unable to get compaction log entries")
+	}
+
+	indexBlobsToDelete := m.findIndexBlobsToDelete(ctx, latestBlob.Timestamp, compactionBlobEntries)
+	compactionLogBlobsToDelete := m.findCompactionLogBlobsToDelete(ctx, latestBlob.Timestamp, compactionBlobs)
```

Comment:
What if something like this occurs -
1. Index blob A is to be deleted and is present as input in log compaction blob L.
2. This function is called after a time sufficiently late to delete both blobs.
3. Index blob A and L are deleted.
4. But in some other snapshot process which just started, index blob A was read (deletes are eventually consistent) while L was not found.

Reply:
fixed in latest commits.